Add option for additional assembly probe dirs

ViktorHofer opened the pull request:
Adds an optional argument to probe for additional assemblies at a different location than the test assembly's directory. This is needed for corefx but should also be useful for advanced scenarios in regular .NET apps.

cc @tonerdo

tonerdo commented:
@ViktorHofer I believe #213 already addresses this?

ViktorHofer commented:
You are absolutely right, I missed that PR. I think both PRs present individual benefits. I will incorporate my changes into @jherby2k's PR.

ViktorHofer commented:
@tonerdo I force pushed @jherby2k and my changes into this PR as I couldn't update his base PR. This includes all the changes. Can you please take a look?

ViktorHofer commented on src/coverlet.core/Coverage.cs:
```diff
@@ -205,29 +218,29 @@ private void CalculateCoverage()
-                    {
-                        document.Value.Branches.Remove(branchToRemove.Key);
-                    }
+                // for MoveNext() compiler autogenerated method we need to patch false positive (IAsyncStateMachine for instance) 
```
These are white-space only changes that are wrong in upstream.

jherby2k commented:
Can you just add https://github.com/ViktorHofer/coverlet/commit/3c480c72edaba748b0a299add502f89e245d8587 to this? I'm so over my head with this complex PR stuff.

ViktorHofer commented:
That is already part of the changes or am I missing something?

jherby2k commented:
As I said I’m completely confused at this point. I added that commit last night to my now fubar’d PR.

I’m not looking at coverlet any more at this point. if this ever makes it into a release, cool.

tonerdo commented:
@jherby2k no worries, your contributions are already included in this PR. I'll do one final sweeping review, your contribution is greatly appreciated

ViktorHofer commented:
@jherby2k your commits are cherry-picked into this PR, you are still the author of the commits and 50% of the work is yours. If I crossed a line please tell me, we usually work this way in our team, therefore I wasn't aware that I could be insensitive.

@tonerdo thanks a lot 👍

jherby2k commented:
No no, its all good. I’m just having a frustrating day in general. Thanks for your contribution!

ViktorHofer commented:
Ok got it. Hope it'll get better soon :)

tonerdo commented on src/coverlet.core/Helpers/InstrumentationHelper.cs:
```diff
@@ -14,11 +14,47 @@ namespace Coverlet.Core.Helpers
+            }
+
+            // The test module's name must be unique.
+            var uniqueModules = new HashSet<string>
```
This has the added benefit of preventing the test module from being included in instrumentation 👍

tonerdo commented on src/coverlet.core/Coverage.cs:
```diff
@@ -40,23 +42,33 @@ public Coverage(string module, string[] excludeFilters, string[] includeFilters,
+                    catch (Exception)
+                    {
+                        InstrumentationHelper.RestoreOriginalModule(module, _identifier);
+                        throw;
```
We shouldn't throw an exception here, instead we should let it continue with other modules. The users will simply see coverage of `0` across board. I'm looking into adding verbose logging that will communicate situations like this to the user

ViktorHofer commented on src/coverlet.core/Helpers/InstrumentationHelper.cs:
```diff
@@ -14,11 +14,47 @@ namespace Coverlet.Core.Helpers
+                moduleDirectory = Directory.GetCurrentDirectory();
+            }
+
+            var dirs = new List<string>(1 + includeDirectories.Length)
```
It has very subtle perf benefits but what we pass in here isn't right as some directories expand later when using the `*`. Therefore I just remove the capacity argument here.

ViktorHofer commented on src/coverlet.core/Helpers/InstrumentationHelper.cs:
```diff
@@ -14,11 +14,47 @@ namespace Coverlet.Core.Helpers
+            foreach (var includeDirectory in includeDirectories.Where(d => d != null))
+            {
+                var fullPath = (!Path.IsPathRooted(includeDirectory)
+                    ? Path.GetFullPath(Path.Combine(moduleDirectory, includeDirectory))
```
That would have caused frustration for customers. Thanks for spotting!

ViktorHofer commented:
@tonerdo Thanks for the through feedback. I addressed all your comments. Hope we can get this into master now.

Additionally I have an ask: Can we please publish a new version to either an internal feed (myget) or nuget (preferably nuget) after this is in? Thanks a lot!

tonerdo commented on src/coverlet.core/Helpers/InstrumentationHelper.cs:
```diff
@@ -40,10 +78,11 @@ public static bool HasPdb(string module)
         }
 
-        public static void BackupOriginalModule(string module, string identifier)
+        public static string BackupOriginalModule(string module, string identifier)
```
Oh! One more question here. Why did you update this to return a string as I don't see it used anywhere

ViktorHofer replied:
@jherby2k changed that to simplify testing: https://github.com/tonerdo/coverlet/pull/253/files#diff-009f00622e86f8b17a0f74fc89bd66a6R31

ViktorHofer replied:
I'm open to revert the change if you wish so.

tonerdo replied:
Please do, thanks!

ViktorHofer replied:
Done. I also added the `overwrite: true` parameter as in local testing when debugging I hit that once. I hope that's fine.

tonerdo replied:
That's fine

tonerdo commented:
Sure @ViktorHofer, will merge and publish once I get an answer to my little nit above and the CI passes

tonerdo commented:
> consider using Azure DevOps

Hehehe, shameless plug 😄

tonerdo commented:
@ViktorHofer there's a new release with these changes

ViktorHofer commented:
Amazing, thanks 😊

jherby2k commented:
Thanks fellas!
